core/chains/chains.go

```diff
@@ -0,0 +1,131 @@
+package chains
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+)
+
+// To add a new chain to this file:
+// 1. Add the global var in the vars list
+// 2. Add the chain ID in the map in the init() function
+// 3. Add a config set in configs.go
+
+// Chain represents a blockchain with a unique Chain ID
+type Chain struct {
+	id      *big.Int
+	config  ChainSpecificConfig
+	logOnce sync.Once
+}
+
+func (c *Chain) setChainID(id int64) {
+	c.id = big.NewInt(id)
+}
+
+func (c *Chain) ID() *big.Int {
+	return c.id
+}
+
+func (c *Chain) Config() ChainSpecificConfig {
+	if !c.config.set {
+		c.logOnce.Do(func() {
+			logger.Warnf("chain with ID %s does not have a chain-specific config, using fallback config instead", c.ID())
+		})
+		return FallbackConfig
+	}
+	return c.config
+}
+
+// IsArbitrum returns true if the chain is arbitrum mainnet or testnet
+func (c *Chain) IsArbitrum() bool {
+	if c == nil {
+		return false
+	}
+	return c.ID().Cmp(ArbitrumMainnet.ID()) == 0 || c.ID().Cmp(ArbitrumRinkeby.ID()) == 0
+}
+
+// IsOptimism returns true if the chain is optimism mainnet or testnet
+func (c *Chain) IsOptimism() bool {
+	if c == nil {
+		return false
+	}
+	return c.ID().Cmp(OptimismMainnet.ID()) == 0 || c.ID().Cmp(OptimismKovan.ID()) == 0
+}
+
+var chains map[int64]*Chain
+var (
+	EthMainnet      = new(Chain)
+	EthRinkeby      = new(Chain)
+	EthGoerli       = new(Chain)
+	EthKovan        = new(Chain)
+	OptimismMainnet = new(Chain)
+	OptimismKovan   = new(Chain)
+	ArbitrumMainnet = new(Chain)
+	ArbitrumRinkeby = new(Chain)
+	BSCMainnet      = new(Chain)
+	HecoMainnet     = new(Chain)
+	FantomMainnet   = new(Chain)
+	FantomTestnet   = new(Chain)
+	PolygonMainnet  = new(Chain)
+	PolygonMumbai   = new(Chain)
+	XDaiMainnet     = new(Chain)
+	RSKMainnet      = new(Chain)
+	AvalancheFuji   = new(Chain)
+)
+
+func init() {
+	chains = make(map[int64]*Chain)
+
+	chains[1] = EthMainnet
+	chains[4] = EthRinkeby
+	chains[5] = EthGoerli
+	chains[42] = EthKovan
+	chains[10] = OptimismMainnet
+	chains[69] = OptimismKovan
+	chains[42161] = ArbitrumMainnet
+	chains[421611] = ArbitrumRinkeby
+	chains[56] = BSCMainnet
+	chains[128] = HecoMainnet
+	chains[250] = FantomMainnet
+	chains[4002] = FantomTestnet
+	chains[137] = PolygonMainnet
+	chains[80001] = PolygonMumbai
+	chains[100] = XDaiMainnet
+	chains[30] = RSKMainnet
+	chains[43113] = AvalancheFuji
+
+	for id, chain := range chains {
+		chain.setChainID(id)
+	}
+
+	setConfigs()
+}
+
+var chainsMu sync.RWMutex
+
+// ChainFromID returns the chain for the given ID
+// If no chain is found, creates a new one and returns that
+func ChainFromID(id *big.Int) *Chain {
+	if !id.IsInt64() {
+		panic("chain IDs larger than the max 64 bit integer are not currently supported")
+	}
+	// Use rlock for the hot path
+	chainsMu.RLock()
```

Comment:
The amount of locking and unlocking in this functions seems a bit like premature optimization? A single Lock + defer Unlock seems easier to reason about IMHO.

Reply:
I wanted to be careful since we might query for `Chain` quite a bit but you are right it is probably premature optimisation.